lib/dpl/provider/pages.rb

```diff
@@ -44,34 +53,140 @@ def slug
         options.fetch(:repo) { context.env['TRAVIS_REPO_SLUG'] }
       end
 
+      def keep_history
+        options.fetch(:keep_history) { false }
+      end
+
+      def allow_empty_commit
+        options.fetch(:allow_empty_commit) { false }
+      end
+
+      def verbose
+        # Achtung! Never verbosify git, since it may expose user's token.
+        options.fetch(:verbose) { false }
+      end
+
       def check_auth
       end
 
       def needs_key?
         false
       end
 
-      def github_deploy
-        context.shell 'rm -rf .git > /dev/null 2>&1'
-        context.shell "touch \"deployed at `date` by #{@gh_name}\""
-        context.shell 'git init' or raise 'Could not create new git repo'
+      def print_step(msg)
+        $stderr.puts msg if @verbose
+      end
+
+      def github_pull(target_dir)
+        print_step "Trying to clone a single branch #{@target_branch} from existing repo..."
+        unless context.shell "git clone --quiet --branch='#{@target_branch}' --single-branch '#{@gh_remote_url}' '#{target_dir}' &>/dev/null"
+          # if such branch doesn't exist at remote, do normal clone and create
+          # a new orphan branch
+          print_step "Cloning #{@target_branch} branch failed"
+          print_step 'Trying to clone the whole repo...'
+          context.shell "git clone --quiet '#{@gh_remote_url}' '#{target_dir}' &>/dev/null" or raise "It looks the repo doesn't exist on remote or is inaccessible"
+          FileUtils.cd(target_dir, :verbose => @verbose) do
+            print_step "Assuming #{@target_branch} branch doesn't exist, thus creating orphan one"
+            context.shell "git checkout --orphan '#{@target_branch}'"
+          end
+        end
+      end
+
+      def github_clean
+        context.shell "pwd"
+        print_step 'Purging every existing file from repo...'
+        context.shell "git ls-files -z 2>/dev/null | xargs -0 rm -f 2>/dev/null"  # remove all committed files from the repo
```

Comment:
It's this complexity that `rsync --exclude .git --delete` avoids. Those arguments tell `rsync` to ignore the `.git` directory and to remove everything else that's not in the source, which should match the behavior you've hand-coded here.

Reply:
Sounds interesting. I was used to consider git porcelain more reliable, but looks you are right.